Framework/CurveFitting/src/Functions/ChudleyElliotSQE.cpp

```diff
@@ -0,0 +1,118 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+// Mantid Coding standars <http://www.mantidproject.org/Coding_Standards>
+// Main Module Header
+#include "MantidCurveFitting/Functions/ChudleyElliotSQE.h"
+// Mantid Headers from the same project
+// N/A
+// Mantid headers from other projects
+#include "MantidAPI/FunctionFactory.h"
+#include "MantidAPI/IFunction.h"
+#include "MantidAPI/Jacobian.h"
+// third party library headers
+// N/A
+// standard library headers
+#include <cmath>
+#include <limits>
+
+namespace {
+Mantid::Kernel::Logger g_log("ChudleyElliotSQE");
+}
+
+namespace Mantid::CurveFitting::Functions {
+
+DECLARE_FUNCTION(ChudleyElliotSQE)
+
+/**
+ * @brief function parameters and impose constraints
+ */
+void ChudleyElliotSQE::declareParameters() {
+  this->declareParameter("Height", 1.0, "scaling factor");
+  this->declareParameter("Centre", 0.0, "Shift along the X-axis");
+  this->declareParameter("Tau", 1.25, "Residence time");
+  this->declareParameter("L", 1.25, "Jump length");
+}
+
+/**
+ * @brief Calculate function values on an energy domain
+ * @param out array to store function values
+ * @param xValues energy domain where function is evaluated
+ * @param nData size of the energy domain
+ */
+void ChudleyElliotSQE::function1D(double *out, const double *xValues, const size_t nData) const {
+  double hbar(0.658211626); // ps*meV
+  auto H = this->getParameter("Height");
+  auto C = this->getParameter("Centre");
+  auto Q = this->getAttribute("Q").asDouble();
+  auto T = this->getParameter("Tau");
+  auto L = this->getParameter("L");
+
+  // Penalize negative parameters, just in case they show up
+  // when calculating the numeric derivative
+  if (H < std::numeric_limits<double>::epsilon() || T < std::numeric_limits<double>::epsilon()) {
+    for (size_t j = 0; j < nData; j++) {
+      out[j] = std::numeric_limits<double>::infinity();
+    }
+    return;
+  }
+
+  // Lorentzian intensities and HWHM
+  auto G = hbar * (1 - (sin(Q * L) / (Q * L))) / T;
+  for (size_t j = 0; j < nData; j++) {
+    auto E = xValues[j] - C;
+    out[j] += H * G / (G * G + E * E) / M_PI;
+  }
+}
+
+/**
+ * @brief analytical/numerical derivative with respect to fitting parameters
+ * We carry out analytical derivative for parameter Height and numerical
+ * derivatives for parameters
+ * DiffCoeff, Tau, and Centre, selecting sensible steps.
```

Comment:
Would not it be better to include below the analytical derivatives of this function with respect to parameters "Centre", "Tau" and "L" in order not to limit a user to a specific range of values where the numerical approximation is feasible? It does not seem to be much work, as the symbolic derivatives can be handled with the help of an external software/library (e.g. python SymPy library)

Reply:
Yes. The simpler versions for FqFit use analytic derivatives.

Reply:
Would this be better done in an upgrade Issue?

Reply:
Sounds good to me